File: code/frontend/scripts/export-presence.mjs
import { createHash } from 'node:crypto'
import { readdir, readFile, mkdir, writeFile, rename, access } from 'node:fs/promises'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { HOUR, dayOf, circularBounds } from '../src/presence/timeline.mjs'

const frontend = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..')
const hash = value => createHash('sha256').update(typeof value === 'string' ? value : JSON.stringify(value)).digest('hex')
const text = value => value == null || String(value).trim() === '' ? null : String(value).trim()
const canonical = value => Array.isArray(value) ? value.map(canonical) : value && typeof value === 'object' ? Object.fromEntries(Object.keys(value).sort().map(key => [key, canonical(value[key])])) : value

// Preserve original numeric tokens (including Python's 1.0 and 1e-07),
// normalize string escaping, sort object keys, and discard whitespace.
// JSON.parse validates the input before this canonical integrity check.
function sourceCanonicalJSON(raw) {
  const tokens = raw.match(/"(?:\\.|[^"\\])*"|-?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?|true|false|null|[{}\[\]:,]/g)
  let index = 0
  const quote = value => JSON.stringify(value).replace(/[\u007f-\uffff]/g, char => `\\u${char.charCodeAt(0).toString(16).padStart(4, '0')}`)
  function value() {
    const token = tokens[index++]
    if (token === '{') {
      const pairs = []
      while (tokens[index] !== '}') {
        const key = JSON.parse(tokens[index++]); index++
        pairs.push([key, value()])
        if (tokens[index] === ',') index++
      }
      index++; return `{${pairs.sort((a, b) => a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0).map(([key, entry]) => `${quote(key)}:${entry}`).join(',')}}`
    }
    if (token === '[') {
      const entries = []
      while (tokens[index] !== ']') { entries.push(value()); if (tokens[index] === ',') index++ }
      index++; return `[${entries.join(',')}]`
    }
    return token.startsWith('"') ? quote(JSON.parse(token)) : token
  }
  return value()
}

async function discover(directory) {
  const found = []
  for (const entry of (await readdir(directory, { withFileTypes: true })).sort((a, b) => a.name.localeCompare(b.name))) {
    const target = path.join(directory, entry.name)
    if (entry.isDirectory()) found.push(...await discover(target))
    else if (entry.name === 'report.json') found.push(target)
  }
  return found
}

function assert(condition, message) { if (!condition) throw new Error(message) }
function utcHour(value) {
  assert(typeof value === 'string' && /^\d{4}-\d\d-\d\d[T ]\d\d:\d\d(?::00(?:\.000)?)?(?:Z)?$/.test(value), `Invalid UTC hour: ${value}`)
  const normalized = value.replace(' ', 'T').replace(/Z$/, '')
  const ts = Date.parse(`${normalized}Z`)
  assert(Number.isFinite(ts) && ts % HOUR === 0 && new Date(ts).toISOString().slice(0, 16) === normalized.slice(0, 16), `Invalid UTC hour: ${value}`)
  return ts
}

function readRows(payload, expectedDataset) {
  assert(payload && Array.isArray(payload.entries), 'Missing report entries array')
  assert(payload.nextOffset == null && (payload.offset == null || payload.offset === 0), 'Incomplete paginated report')
  // 4Wings total counts dataset envelopes, not observation rows.
  assert(payload.total == null || payload.total === payload.entries.length, 'Report envelope total does not match entries')
  const rows = []
  for (const entry of payload.entries) {
    assert(entry && typeof entry === 'object' && !Array.isArray(entry), 'Invalid dataset envelope')
    assert(Object.keys(entry).length > 0, 'Empty dataset envelope')
    for (const [datasetVersion, values] of Object.entries(entry)) {
      assert(datasetVersion.startsWith('public-global-presence:') && Array.isArray(values), 'Unsupported presence dataset envelope')
      assert(datasetVersion === expectedDataset, 'Dataset version differs from manifest')
      for (const row of values) { assert(row && typeof row === 'object' && !Array.isArray(row), 'Invalid observation row'); rows.push({ datasetVersion, row }) }
    }
  }
  return rows
}

export async function exportPresence({ input = path.resolve(frontend, '../../data/bronze/gfw_presence'), output = path.join(frontend, 'public/data/presence') } = {}) {
  const files = await discover(input)
  assert(files.length > 0, `No presence report.json files found in ${input}`)
  const observations = new Map(), coverages = new Map(), parsedPayloads = new Map()
  for (const file of files) {
    try {
      const raw = await readFile(file, 'utf8')
      const payload = JSON.parse(raw)
      const manifest = JSON.parse(await readFile(path.join(path.dirname(file), 'manifest.json'), 'utf8'))
      const request = manifest.request ?? {}
      assert(request['temporal-resolution'] === 'HOURLY' && request['spatial-aggregation'] === false && request['group-by'] === 'VESSEL_ID', 'Requires unaggregated HOURLY VESSEL_ID presence report')
      const gridResolution = { HIGH: 0.01, LOW: 0.1 }[request['spatial-resolution']]
      assert(gridResolution, 'Unsupported spatial resolution')
      assert(typeof request['region-dataset'] === 'string' && Number.isInteger(request['region-id']), 'Missing region identity')
      const range = request['date-range']?.split(',')
      assert(range?.length === 2, 'Missing requested date range')
      const start = utcHour(range[0]), end = utcHour(range[1])
      assert(start < end, 'Requested range must have start before end')
      assert(typeof manifest.dataset_version === 'string' && manifest.dataset_version.startsWith('public-global-presence:'), 'Missing presence dataset version')
      const updated = Date.parse(manifest.created_at)
      assert(Number.isFinite(updated), 'Missing valid retrieval creation time')
      const rank = `${new Date(updated).toISOString()}|${path.relative(input, file)}`
      // Compute our own canonical fingerprint. Upstream hashes use Python JSON
      // number formatting and must not be compared to raw file bytes or JS JSON.
      const fingerprint = hash(canonical(payload))
      if (manifest.raw_response_sha256 != null) {
        assert(/^[a-f0-9]{64}$/.test(manifest.raw_response_sha256), 'Invalid manifest payload hash')
        assert(manifest.raw_response_sha256 === fingerprint || manifest.raw_response_sha256 === hash(sourceCanonicalJSON(raw)), 'Manifest payload hash does not match report contents')
      }
      let rows = parsedPayloads.get(fingerprint)
      if (!rows) { rows = readRows(payload, manifest.dataset_version); parsedPayloads.set(fingerprint, rows) }
      else readRows(payload, manifest.dataset_version)
      assert(Number.isInteger(manifest.row_count) && manifest.row_count === rows.length, 'Manifest row_count does not match report')
      assert(manifest.valid_position_count == null || manifest.valid_position_count === rows.length, 'Manifest reports invalid positions')
      const coverage = { start: new Date(start).toISOString(), end: new Date(end).toISOString(), regionDataset: request['region-dataset'], regionId: request['region-id'], datasetVersion: manifest.dataset_version, gridResolution }
      coverages.set(JSON.stringify(coverage), coverage)
      for (const { datasetVersion, row } of rows) {
        const timestamp = utcHour(row.date)
        assert(timestamp >= start && timestamp < end, `Observation ${row.date} is outside requested range`)
        assert(datasetVersion === manifest.dataset_version, 'Dataset version differs from manifest')
        assert(typeof row.vesselId === 'string' && text(row.vesselId), 'Missing GFW vessel identity')
        assert(typeof row.lat === 'number' && Number.isFinite(row.lat) && Math.abs(row.lat) <= 90 && typeof row.lon === 'number' && Number.isFinite(row.lon) && Math.abs(row.lon) <= 180, 'Invalid position coordinates')
        assert(typeof row.hours === 'number' && Number.isFinite(row.hours) && row.hours >= 0 && row.hours <= 1, 'Invalid hourly presence duration')
        const vesselId = `gfw:${text(row.vesselId)}`
        const cellLat = Math.round(row.lat / gridResolution), cellLon = Math.round(row.lon / gridResolution)
        const id = hash([datasetVersion, gridResolution, vesselId, timestamp, cellLat, cellLon])
        const recordRank = `${rank}|${new Date(timestamp).toISOString()}|${hash(canonical(row))}`
        const previous = observations.get(id)
        if (previous && previous.rank >= recordRank) continue
        observations.set(id, {
          rank: recordRank,
          observation: { id, vesselId, ts: new Date(timestamp).toISOString(), lat: row.lat, lon: row.lon, presenceHours: row.hours, datasetVersion, gridResolution },
          vessel: { id: vesselId, name: text(row.shipName), mmsi: text(row.mmsi), imo: text(row.imo), callsign: text(row.callsign), flag: text(row.flag), vesselType: text(row.vesselType), metadataUpdatedAt: new Date(updated).toISOString(), metadataRank: recordRank },
        })
      }
    } catch (error) { throw new Error(`${file}: ${error.message}`, { cause: error }) }
  }

  const days = new Map()
  function day(date) { if (!days.has(date)) days.set(date, { date, covered: new Set(), observations: [], vessels: new Map() }); return days.get(date) }
  const coverage = [...coverages.values()].sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)))
  for (const interval of coverage) for (let ts = Date.parse(interval.start); ts < Date.parse(interval.end); ts += HOUR) day(dayOf(ts)).covered.add(new Date(ts).getUTCHours())
  for (const { observation, vessel, rank } of observations.values()) {
    const target = day(observation.ts.slice(0, 10)); target.observations.push(observation)
    const previous = target.vessels.get(vessel.id)
    target.vessels.set(vessel.id, {
      ...(previous && previous.rank > rank ? previous : { ...vessel, rank }),
      firstObservedAt: previous && previous.firstObservedAt < observation.ts ? previous.firstObservedAt : observation.ts,
      lastObservedAt: previous && previous.lastObservedAt > observation.ts ? previous.lastObservedAt : observation.ts,
      observationCount: (previous?.observationCount ?? 0) + 1,
    })
  }
  const assets = new Map()
  function asset(prefix, value) { const body = JSON.stringify(value); const url = `${prefix}.${hash(body)}.json`; assets.set(url, body); return url }
  const catalogDays = [...days.values()].sort((a, b) => a.date.localeCompare(b.date)).map(target => {
    const rows = target.observations.sort((a, b) => a.ts.localeCompare(b.ts) || a.id.localeCompare(b.id))
    const vessels = [...target.vessels.values()].sort((a, b) => a.id.localeCompare(b.id)).map(({ rank, ...vessel }) => vessel)
    const hourlyCounts = Array(24).fill(0)
    for (const row of rows) hourlyCounts[new Date(row.ts).getUTCHours()]++
    return { date: target.date, observationsUrl: asset(`${target.date}.observations`, { date: target.date, observations: rows }), vesselsUrl: asset(`${target.date}.vessels`, { date: target.date, vessels }),
      bounds: circularBounds(rows), coveredHours: [...target.covered].sort((a, b) => a - b), hourlyCounts, observationCount: rows.length, vesselCount: vessels.length }
  })
  const content = { schemaVersion: 1, source: 'Global Fishing Watch', positionSemantics: 'gfw_presence_grid_center_hourly', coverage, days: catalogDays,
    observationCount: observations.size, coveredHourCount: catalogDays.reduce((sum, entry) => sum + entry.coveredHours.length, 0) }
  const catalog = { ...content, revision: hash(content), generatedAt: new Date().toISOString() }
  await mkdir(output, { recursive: true })
  // Old immutable assets remain valid for tabs still using the previous catalog.
  for (const [filename, body] of assets) {
    const destination = path.join(output, filename)
    try { await access(destination); continue } catch (error) { if (error.code !== 'ENOENT') throw error }
    const temporaryAsset = `${destination}.${process.pid}.${Date.now()}.tmp`
    await writeFile(temporaryAsset, body)
    await rename(temporaryAsset, destination)
  }
  const temporary = path.join(output, `.catalog-${process.pid}-${Date.now()}.tmp`)
  await writeFile(temporary, JSON.stringify(catalog, null, 2))
  await rename(temporary, path.join(output, 'catalog.json'))
  return catalog
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  const args = process.argv.slice(2), options = {}
  for (let i = 0; i < args.length; i += 2) {
    if (!['--input', '--output'].includes(args[i]) || !args[i + 1]) throw new Error('Usage: npm run import:presence -- [--input directory] [--output directory]')
    options[args[i].slice(2)] = path.resolve(args[i + 1])
  }
  try { const result = await exportPresence(options); console.log(`Exported ${result.observationCount.toLocaleString('en-US')} observations across ${result.coveredHourCount} covered hours (${result.days.length} UTC days).`) }
  catch (error) { console.error(error.message); process.exitCode = 1 }
}

File: code/backend/scripts/export-presence.py
#!/usr/bin/env python3
"""Publish browser presence assets directly from normalized GFW Silver Parquet."""

from __future__ import annotations

import argparse
import hashlib
import json
import math
import os
import sys
import tempfile
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterator

try:
    import pyarrow.parquet as pq
except ImportError as error:  # pragma: no cover - depends on the runtime environment
    raise SystemExit("Silver presence export requires pyarrow. Install the project Python dependencies first.") from error


HOUR_MS = 3_600_000
ROOT = Path(__file__).resolve().parents[3]
DEFAULT_INPUT = ROOT / "data" / "silver" / "gfw_presence_hourly"
DEFAULT_OUTPUT = ROOT / "code" / "frontend" / "public" / "data" / "presence"
REQUIRED_COLUMNS = {
    "ts", "vessel_id", "lat", "lon", "presence_hours", "report_dataset",
    "grid_resolution_degrees", "vessel_name", "mmsi", "imo", "callsign",
    "flag_state", "gfw_vessel_type", "position_semantics", "is_valid_position",
}


def fail(message: str) -> None:
    raise ValueError(message)


def sha(value: Any) -> str:
    return hashlib.sha256(json.dumps(value, sort_keys=True, separators=(",", ":"), default=str).encode()).hexdigest()


def clean(value: Any) -> str | None:
    if value is None:
        return None
    value = str(value).strip()
    return value or None


def hour(value: Any) -> datetime:
    if isinstance(value, datetime):
        result = value.astimezone(timezone.utc) if value.tzinfo else value.replace(tzinfo=timezone.utc)
    elif isinstance(value, str):
        raw = value.replace(" ", "T").removesuffix("Z")
        result = datetime.fromisoformat(raw).replace(tzinfo=timezone.utc)
    else:
        fail(f"Invalid UTC hour: {value!r}")
    if result.minute or result.second or result.microsecond:
        fail(f"Invalid UTC hour: {value!r}")
    return result


def iso(value: datetime) -> str:
    return value.strftime("%Y-%m-%dT%H:%M:%S.000Z")


def discover(root: Path) -> list[Path]:
    files = sorted(root.rglob("points.parquet")) if root.exists() else []
    if not files:
        fail(f"No Silver presence points.parquet files found in {root}")
    return files


def request_details(manifest: dict[str, Any]) -> tuple[datetime, datetime, float, dict[str, Any]]:
    request = manifest.get("request") or {}
    if request.get("temporal-resolution") != "HOURLY" or request.get("spatial-aggregation") is not False:
        fail("Requires unaggregated HOURLY Silver presence data")
    resolution = {"HIGH": 0.01, "LOW": 0.1}.get(request.get("spatial-resolution"))
    if resolution is None:
        fail("Unsupported spatial resolution")
    if not isinstance(request.get("region-dataset"), str) or not isinstance(request.get("region-id"), int):
        fail("Missing region identity")
    dates = request.get("date-range", "").split(",")
    if len(dates) != 2:
        fail("Missing requested date range")
    start, end = hour(dates[0]), hour(dates[1])
    if start >= end:
        fail("Requested range must have start before end")
    if not isinstance(manifest.get("dataset_version"), str) or not manifest["dataset_version"].startswith("public-global-presence:"):
        fail("Missing presence dataset version")
    return start, end, resolution, request


def rows(file: Path) -> Iterator[dict[str, Any]]:
    parquet = pq.ParquetFile(file)
    fields = set(parquet.schema_arrow.names)
    missing = REQUIRED_COLUMNS - fields
    if missing:
        fail(f"Missing Silver columns: {', '.join(sorted(missing))}")
    for batch in parquet.iter_batches(columns=sorted(REQUIRED_COLUMNS)):
        yield from batch.to_pylist()


def read_presence(input_root: Path) -> tuple[dict[str, dict[str, Any]], dict[str, dict[str, Any]]]:
    observations: dict[str, dict[str, Any]] = {}
    coverages: dict[str, dict[str, Any]] = {}
    for file in discover(input_root):
        try:
            manifest = json.loads(file.with_name("manifest.json").read_text())
            start, end, resolution, request = request_details(manifest)
            created = datetime.fromisoformat(manifest["created_at"].replace("Z", "+00:00")).astimezone(timezone.utc)
            parquet = pq.ParquetFile(file)
            if manifest.get("row_count") != parquet.metadata.num_rows:
                fail("Manifest row_count does not match Parquet rows")
            if manifest.get("valid_position_count") not in (None, parquet.metadata.num_rows):
                fail("Manifest reports invalid positions")
            coverage = {"start": iso(start), "end": iso(end), "regionDataset": request["region-dataset"], "regionId": request["region-id"], "datasetVersion": manifest["dataset_version"], "gridResolution": resolution}
            coverages[json.dumps(coverage, sort_keys=True)] = coverage
            base_rank = f"{iso(created)}|{file.relative_to(input_root)}"
            for row in rows(file):
                timestamp = hour(row["ts"])
                if not start <= timestamp < end:
                    fail(f"Observation {row['ts']} is outside requested range")
                vessel_id = clean(row["vessel_id"])
                if not vessel_id or not vessel_id.startswith("gfw:"):
                    fail("Missing GFW vessel identity")
                if row["position_semantics"] != "gfw_presence_grid_center_hourly" or row["is_valid_position"] is not True:
                    fail("Invalid Silver position semantics or quality")
                lat, lon, duration = float(row["lat"]), float(row["lon"]), float(row["presence_hours"])
                if not math.isfinite(lat) or not math.isfinite(lon) or not -90 <= lat <= 90 or not -180 <= lon <= 180:
                    fail("Invalid position coordinates")
                if not math.isfinite(duration) or not 0 <= duration <= 1:
                    fail("Invalid hourly presence duration")
                dataset = clean(row["report_dataset"])
                if dataset != manifest["dataset_version"]:
                    fail("Dataset version differs from manifest")
                if not math.isclose(float(row["grid_resolution_degrees"]), resolution, rel_tol=0, abs_tol=1e-12):
                    fail("Grid resolution differs from manifest")
                cell = (round(lat / resolution), round(lon / resolution))
                identity = sha([dataset, resolution, vessel_id, iso(timestamp), *cell])
                rank = f"{base_rank}|{iso(timestamp)}|{sha(row)}"
                if identity in observations and observations[identity]["rank"] >= rank:
                    continue
                observations[identity] = {"rank": rank, "observation": {"id": identity, "vesselId": vessel_id, "ts": iso(timestamp), "lat": lat, "lon": lon, "presenceHours": duration, "datasetVersion": dataset, "gridResolution": resolution}, "vessel": {"id": vessel_id, "name": clean(row["vessel_name"]), "mmsi": clean(row["mmsi"]), "imo": clean(row["imo"]), "callsign": clean(row["callsign"]), "flag": clean(row["flag_state"]), "vesselType": clean(row["gfw_vessel_type"]), "metadataUpdatedAt": iso(created), "metadataRank": rank}}
        except Exception as error:
            raise ValueError(f"{file}: {error}") from error
    return observations, coverages


def write_json(path: Path, value: Any) -> None:
    body = json.dumps(value, separators=(",", ":"), allow_nan=False)
    if path.exists():
        return
    fd, temporary = tempfile.mkstemp(prefix=f".{path.name}.", suffix=".tmp", dir=path.parent)
    with os.fdopen(fd, "w") as stream:
        stream.write(body)
    os.replace(temporary, path)


def circular_bounds(observations: list[dict[str, Any]]) -> list[float] | None:
    """Return a compact longitude interval; crossing the dateline is valid."""
    if not observations:
        return None
    lats = [item["lat"] for item in observations]
    longitudes = sorted((item["lon"] + 360) % 360 for item in observations)
    if len(longitudes) == 1:
        west = east = longitudes[0]
    else:
        gaps = [(longitudes[(index + 1) % len(longitudes)] - value) % 360 for index, value in enumerate(longitudes)]
        cut = gaps.index(max(gaps))
        west, east = longitudes[(cut + 1) % len(longitudes)], longitudes[cut]
    normalize = lambda value: value - 360 if value > 180 else value
    return [normalize(west), min(lats), normalize(east), max(lats)]


def export(input_root: Path, output: Path) -> dict[str, Any]:
    observations, coverages = read_presence(input_root)
    days: dict[str, dict[str, Any]] = {}
    def day(date: str) -> dict[str, Any]:
        return days.setdefault(date, {"date": date, "covered": set(), "observations": [], "vessels": {}})
    for coverage in coverages.values():
        cursor, end = hour(coverage["start"]), hour(coverage["end"])
        while cursor < end:
            day(cursor.date().isoformat())["covered"].add(cursor.hour); cursor = datetime.fromtimestamp(cursor.timestamp() + 3600, timezone.utc)
    for item in observations.values():
        observation, vessel, rank = item["observation"], item["vessel"], item["rank"]
        target = day(observation["ts"][:10]); target["observations"].append(observation)
        previous = target["vessels"].get(vessel["id"])
        chosen = previous if previous and previous["rank"] > rank else {**vessel, "rank": rank}
        target["vessels"][vessel["id"]] = {**chosen, "firstObservedAt": min(previous["firstObservedAt"], observation["ts"]) if previous else observation["ts"], "lastObservedAt": max(previous["lastObservedAt"], observation["ts"]) if previous else observation["ts"], "observationCount": (previous["observationCount"] if previous else 0) + 1}
    output.mkdir(parents=True, exist_ok=True); catalog_days = []
    for date, target in sorted(days.items()):
        values = sorted(target["observations"], key=lambda entry: (entry["ts"], entry["id"])); counts = [0] * 24
        for value in values: counts[hour(value["ts"]).hour] += 1
        vessels = [{key: value for key, value in entry.items() if key != "rank"} for _, entry in sorted(target["vessels"].items())]
        observation_payload, vessel_payload = {"date": date, "observations": values}, {"date": date, "vessels": vessels}
        observations_url = f"{date}.observations.{sha(observation_payload)}.json"
        vessels_url = f"{date}.vessels.{sha(vessel_payload)}.json"
        write_json(output / observations_url, observation_payload); write_json(output / vessels_url, vessel_payload)
        catalog_days.append({"date": date, "observationsUrl": observations_url, "vesselsUrl": vessels_url, "bounds": circular_bounds(values), "coveredHours": sorted(target["covered"]), "hourlyCounts": counts, "observationCount": len(values), "vesselCount": len(vessels)})
    content = {"schemaVersion": 1, "source": "Global Fishing Watch", "positionSemantics": "gfw_presence_grid_center_hourly", "coverage": sorted(coverages.values(), key=lambda entry: json.dumps(entry, sort_keys=True)), "days": catalog_days, "observationCount": len(observations), "coveredHourCount": sum(len(entry["coveredHours"]) for entry in catalog_days)}
    catalog = {**content, "revision": sha(content), "generatedAt": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.000Z")}
    temporary = output / f".catalog-{os.getpid()}.tmp"; temporary.write_text(json.dumps(catalog, indent=2)); os.replace(temporary, output / "catalog.json")
    return catalog


if __name__ == "__main__":
    parser = argparse.ArgumentParser(); parser.add_argument("--input", type=Path, default=DEFAULT_INPUT); parser.add_argument("--output", type=Path, default=DEFAULT_OUTPUT)
    args = parser.parse_args()
    try:
        result = export(args.input.resolve(), args.output.resolve())
        print(f"Exported {result['observationCount']:,} observations across {result['coveredHourCount']} covered hours ({len(result['days'])} UTC days).")
    except Exception as error:
        print(error, file=sys.stderr); raise SystemExit(1)
